File: bench/fannkuchredux/fannkuchredux.scala
/* The Computer Language Benchmarks Game
   http://shootout.alioth.debian.org/

   conversion to Scala by Rex Kerr
   from Java version by Oleg Mazurov and Isaac Gouy
*/

object fannkuchredux
{
  def fannkuch(n: Int): Int = {
    val perm1 = Array.range(0,n)
    val perm, count = new Array[Int](n)
    var f, i, k, r, flips, nperm, checksum = 0

    r = n
    while (r>0) {
      i = 0
      while (r != 1) { count(r-1) = r; r -= 1 }
      while (i < n) { perm(i) = perm1(i); i += 1 }
      
      // Count flips and update max  and checksum
      f = 0
      k = perm(0)
      while (k != 0) {
        i = 0
        while (2*i < k) {
          val t = perm(i); perm(i) = perm(k-i); perm(k-i) = t
          i += 1
        }
        k = perm(0)
        f += 1
      }
      if (f>flips) flips = f
      if ((nperm&0x1)==0) checksum += f
      else checksum -= f

      // Use incremental change to generate another permutation
      var go = true
      while (go) {
        if (r == n) {
          println(checksum)
          return flips
        }
        val p0 = perm1(0)
        i = 0
        while (i < r) {
          val j = i+1
          perm1(i) = perm1(j)
          i = j
        }
        perm1(r) = p0

        count(r) -= 1
        if (count(r) > 0) go = false
        else r += 1
      }
      nperm += 1
    }
    flips
  }

  def main(args: Array[String]) {
    val n = (if (args.length > 0) args(0).toInt else 7)
    println("Pfannkuchen("+n+") = "+fannkuch(n))
  }
}

File: bench/fannkuchredux/fannkuchredux.go
/*
 * The Computer Language Benchmarks Game
 * http://shootout.alioth.debian.org/
 *
 * contributed by Oleg Mazurov, June 2010
 *
 */

package main

import (
    "fmt"
    "os"
    "runtime"
    "strconv"
)

type Result struct {
    maxFlips int
    checkSum int
}

var (
    NCHUNKS = 720
    CHUNKSZ = 0
    NTASKS  = 0
)
var n = 12
var Fact []int

func fannkuch( idxMin int, ch chan Result ) {

    idxMax := idxMin + CHUNKSZ
    if idxMax < Fact[n] {
        go fannkuch( idxMax, ch )
    } else {
        idxMax = Fact[n]
    }

    p     := make([]int, n)
    pp    := make([]int, n)
    count := make([]int, n)

    // first permutation
    for i := 0; i<n; i++ {
        p[i] = i
    }
    for i, idx := n-1, idxMin; i>0; i-- {
        d := idx / Fact[i]
        count[i] = d
        idx = idx % Fact[i]

        copy( pp, p )
        for j := 0; j <= i; j++ {
	    if j+d <= i {
                p[j] = pp[j+d]
	    } else {
                p[j] = pp[j+d-i-1]
	    }
        }
    }

    maxFlips := 1
    checkSum := 0

    for idx, sign := idxMin, true; ; sign = !sign {

        // count flips
        first := p[0]
	if first != 0 {
	    flips := 1
	    if p[first] != 0 {
		copy( pp, p )
		p0 := first
	        for {
		    flips++
		    for i, j := 1, p0-1; i < j; i, j = i+1, j-1 {
		        pp[i], pp[j] = pp[j], pp[i]
		    }
		    t := pp[p0]
		    pp[p0] = p0
		    p0 = t
		    if pp[p0] == 0 {
		        break
		    }
	        }
	    }
	    if maxFlips < flips {
		maxFlips = flips
	    }
	    if sign {
		checkSum += flips
	    } else {
		checkSum -= flips
	    }
	}

	if idx++; idx == idxMax {
	    break
	}

	// next permutation
	if sign {
	    p[0], p[1] = p[1], first
	} else {
	    p[1], p[2] = p[2], p[1]
	    for k := 2;; k++ {
	        if count[k]++; count[k] <= k {
		    break
		}
	        count[k] = 0
		for j:=0; j<=k; j++ {
		    p[j] = p[j+1]
		}
		p[k+1] = first
		first = p[0]
	    }
	}
    }

    ch <- Result{ maxFlips, checkSum }
}

func printResult( n int, res int, chk int ) {
    fmt.Printf("%d\nPfannkuchen(%d) = %d\n", chk, n, res)
}

func main() {

    if len(os.Args) > 1 {
	v, err := strconv.Atoi(os.Args[1])
	if err != nil {
	    fmt.Printf( "error: %s\n", err.String() )
	    return
	}
	n = v
    }
    if n <= 0 || n > 12 {
	printResult( n, -1, -1 )
	return
    }

    runtime.GOMAXPROCS(4)

    Fact = make([]int, n+1)
    Fact[0] = 1
    for i := 1; i<len(Fact); i++ {
        Fact[i] = Fact[i-1] * i
    }

    CHUNKSZ = (Fact[n] + NCHUNKS - 1) / NCHUNKS
    CHUNKSZ += CHUNKSZ%2
    NTASKS = (Fact[n] + CHUNKSZ - 1) / CHUNKSZ

    ch := make(chan Result, NTASKS)

    go fannkuch(0, ch)
    
    res := 0
    chk := 0
    for i := 0; i<NTASKS; i++ {
	r := <-ch
	if res < r.maxFlips {
            res = r.maxFlips
	}
	chk += r.checkSum
    }

    printResult( n, res, chk )
}
